rust/connlib/clients/android/src/make_writer.rs

```diff
@@ -0,0 +1,88 @@
+//! Heavily inspired from https://github.com/Actyx/tracing-android/blob/master/src/android.rs.
+
+use std::{
+    ffi::{CStr, CString},
+    io::{self, BufWriter},
+};
+use tracing::Level;
+
+const LOGGING_MSG_MAX_LEN: usize = 4000;
+
+pub(crate) struct MakeWriter {
+    tag: CString,
+}
+
+pub(crate) struct Writer {
+    level: Level,
+    tag: CString,
+}
+
+impl MakeWriter {
+    pub(crate) fn new(tag: &'static str) -> Self {
+        Self {
+            tag: CString::new(tag).expect("tag must not contain nul-byte"),
+        }
+    }
+
+    fn make_writer_for_level(&self, level: Level) -> BufWriter<Writer> {
+        let inner = Writer {
+            level,
+            tag: self.tag.clone(),
+        };
+
+        BufWriter::with_capacity(LOGGING_MSG_MAX_LEN, inner)
+    }
+}
+
+impl tracing_subscriber::fmt::MakeWriter<'_> for MakeWriter {
+    type Writer = BufWriter<Writer>;
+
+    fn make_writer(&self) -> Self::Writer {
+        self.make_writer_for_level(Level::INFO)
+    }
+
+    fn make_writer_for(&self, meta: &tracing::Metadata<'_>) -> Self::Writer {
+        self.make_writer_for_level(*meta.level())
+    }
+}
+
+impl io::Write for Writer {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        let written = buf.len().min(LOGGING_MSG_MAX_LEN);
+
+        let msg = &buf[..written];
+        let msg = CString::new(msg.to_vec())?;
```

Comment:
Theory: Perhaps this is failing because we are now constructing a more "complex" message that can't be turned into a `CString`?